In [0]:
import pandas as pd
import numpy as np
from google.colab import drive
import seaborn as sn
import matplotlib.pyplot as plt
from sklearn import preprocessing
from sklearn.preprocessing import OneHotEncoder
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import LabelEncoder 
import os

***Week1-4***

In [3]:

drive.mount('/content/drive', force_remount=True)

Mounted at /content/drive


In [0]:
os.chdir("/content/drive/My Drive")

In [0]:
transaction_train = pd.read_csv('train_transaction.csv')
identity_train = pd.read_csv('train_identity.csv')

***A) Data preprocessing***

Let's join transaction_train and identity_train.

In [0]:
dataset = transaction_train.join(identity_train.set_index('TransactionID'), on= 'TransactionID')  

In [10]:
dataset.shape, transaction_train.shape, identity_train.shape

((590540, 434), (590540, 394), (144233, 41))

We take a look on the head

In [11]:
dataset.head(5)

TransactionID  isFraud  ...  DeviceType                     DeviceInfo
0        2987000        0  ...         NaN                            NaN
1        2987001        0  ...         NaN                            NaN
2        2987002        0  ...         NaN                            NaN
3        2987003        0  ...         NaN                            NaN
4        2987004        0  ...      mobile  SAMSUNG SM-G892A Build/NRD90M

[5 rows x 434 columns]

 We don't know what each column describes exactly. After looking on kaggle website discussion, we found the following information.

TransactionDT: timedelta from a given reference datetime (not an actual timestamp)

TransactionAMT : That is transaction amount in USD

ProductCD: Product code for each transaction

card1 - card6: Information about payment card 

addr: Address for purchaser and recipient

addr1 as billing region

addr2 as billing country

dist: distance
"distances between (not limited) billing address, mailing address, zip code, IP address, phone area, etc.


P_ and (R_) emaildomain: Purchaser and recipient email domain 

C1-C14: Counting, such as how many addresses are found to be associated with the payment card, etc. The actual meaning is masked.

D1-D15: Timedelta, such as days between previous transaction, etc.

M1-M9: Match, such as names on card and address, etc.

Vxxx: Vesta engineered rich features, including ranking, counting, and other entity relations.

id01-id11 : Numerical features for identity

IsFraud =1 means fradulent transaction, otherwise nofradulent transaction. 


Let us discover the number of column and row

In [12]:
print("The shape of data :",dataset.shape)

The shape of data : (590540, 434)


We have 434 features with 590540 rows.
What are columns names?

Let's check missing value.

In [13]:
dataset.isnull().sum()

TransactionID          0
isFraud                0
TransactionDT          0
TransactionAmt         0
ProductCD              0
                   ...  
id_36             449555
id_37             449555
id_38             449555
DeviceType        449730
DeviceInfo        471874
Length: 434, dtype: int64

We decide to remove features with 75% of NaN in their observations.

In [0]:
for i in dataset.columns:
  percent = dataset[i].isnull().sum()/dataset.shape[0]
  
  if percent >= 0.75 :
    dataset = dataset.drop([i], axis = 1)

Let's extract the target variable (IsFraud).

In [0]:
y = dataset.isFraud 

In [16]:
y.value_counts(normalize= True)

0    0.96501
1    0.03499
Name: isFraud, dtype: float64

We have 96% of non fradulent transaction and 4% of non fradulent transaction.

Now, we drop isFraud and TransactionID in the dataset.

In [0]:
dataset = dataset.drop(['isFraud','TransactionID'], axis = 1 )

In [18]:
print("The shape of trainig data after dropping features:", dataset.shape)

The shape of trainig data after dropping features: (590540, 224)


 Data cleaning

Let's replace missing values. We will use datacleaner library for dealing with numerical features. The function autoclean from datacleaner library replaces Nan of continuous features by its median.

Its documentation can be find there 
https://github.com/rhiever/datacleaner

In [19]:
!pip install datacleaner

In [0]:
from datacleaner import autoclean

We create a mask for categorical features

In [21]:
mask1 =dataset.dtypes==object
categoricalColumns = dataset.columns[mask1]
categoricalFeatures = dataset[categoricalColumns]
categoricalFeatures.head()

ProductCD       card4   card6 P_emaildomain   M1  ...   M5   M6   M7   M8   M9
0         W    discover  credit           NaN    T  ...    F    T  NaN  NaN  NaN
1         W  mastercard  credit     gmail.com  NaN  ...    T    T  NaN  NaN  NaN
2         W        visa   debit   outlook.com    T  ...    F    F    F    F    F
3         W  mastercard   debit     yahoo.com  NaN  ...    T    F  NaN  NaN  NaN
4         H  mastercard  credit     gmail.com  NaN  ...  NaN  NaN  NaN  NaN  NaN

[5 rows x 13 columns]

We create a mask for numerical features

In [22]:
mask2 =dataset.dtypes!=object
numericalColumns = dataset.columns[mask2]
numericalFeatures = dataset[numericalColumns]
numericalFeatures.head()

TransactionDT  TransactionAmt  card1  card2  ...   V318  V319  V320  V321
0          86400            68.5  13926    NaN  ...    0.0   0.0   0.0   0.0
1          86401            29.0   2755  404.0  ...    0.0   0.0   0.0   0.0
2          86469            59.0   4663  490.0  ...    0.0   0.0   0.0   0.0
3          86499            50.0  18132  567.0  ...  790.0   0.0   0.0   0.0
4          86506            50.0   4497  514.0  ...    0.0   0.0   0.0   0.0

[5 rows x 211 columns]

In [23]:
numericalFeatures = autoclean(numericalFeatures)

/usr/local/lib/python3.6/dist-packages/pandas/core/generic.py:6287: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._update_inplace(new_data)


In [24]:
numericalFeatures.head()

TransactionDT  TransactionAmt  card1  card2  ...   V318  V319  V320  V321
0          86400            68.5  13926  361.0  ...    0.0   0.0   0.0   0.0
1          86401            29.0   2755  404.0  ...    0.0   0.0   0.0   0.0
2          86469            59.0   4663  490.0  ...    0.0   0.0   0.0   0.0
3          86499            50.0  18132  567.0  ...  790.0   0.0   0.0   0.0
4          86506            50.0   4497  514.0  ...    0.0   0.0   0.0   0.0

[5 rows x 211 columns]

Standardization of numerical features.

In [0]:
numericalFeatures=(numericalFeatures-numericalFeatures.mean())/numericalFeatures.std()

In [27]:
numericalFeatures.head()

TransactionDT  TransactionAmt     card1  ...      V319      V320      V321
0      -1.577985       -0.278167  0.821695  ... -0.055287 -0.088855 -0.074142
1      -1.577985       -0.443327 -1.457557  ... -0.055287 -0.088855 -0.074142
2      -1.577970       -0.317889 -1.068262  ... -0.055287 -0.088855 -0.074142
3      -1.577964       -0.355520  1.679857  ... -0.055287 -0.088855 -0.074142
4      -1.577962       -0.355520 -1.102132  ... -0.055287 -0.088855 -0.074142

[5 rows x 211 columns]

In [0]:
def encoding():
  x = pd.get_dummies(data=categoricalFeatures, columns= categoricalFeatures.columns)
  encoded = numericalFeatures.join(x)
  return encoded
  

In [0]:
datasetEncoded = encoding()

In [30]:
datasetEncoded.shape

(590540, 302)

In [31]:
datasetEncoded.head()

TransactionDT  TransactionAmt     card1     card2  ...  M8_F  M8_T  M9_F  M9_T
0      -1.577985       -0.278167  0.821695 -0.009783  ...     0     0     0     0
1      -1.577985       -0.443327 -1.457557  0.264810  ...     0     0     0     0
2      -1.577970       -0.317889 -1.068262  0.813996  ...     1     0     1     0
3      -1.577964       -0.355520  1.679857  1.305709  ...     0     0     0     0
4      -1.577962       -0.355520 -1.102132  0.967258  ...     0     0     0     0

[5 rows x 302 columns]

***Week5-Today***

***B) Training and prediction***

Let's define auc function which return auc_score.

In [0]:
from sklearn.metrics import roc_auc_score
def auc(y_true, y_pred):
    """
        return the AUC
        ex : auc(y_true, y_prediction)
    """
    return roc_auc_score(y_true, y_pred)

***Hyperparamaters tunning for Knn***

In [0]:
from sklearn.model_selection import GridSearchCV
from sklearn.model_selection import cross_val_score
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import train_test_split

In [0]:
p = 0.3
X_train, X_test, y_train, y_test = train_test_split( datasetEncoded, y, test_size = p, random_state = 42)

In [0]:
def auc(y_true, y_pred):
    """
        return the AUC
        ex : auc(y_true, y_prediction)
    """
    return roc_auc_score(y_true, y_pred)

In [0]:
model = KNeighborsClassifier()
parameters = {'n_neighbors': [2,3,5]} 

In [0]:
from sklearn.model_selection import GridSearchCV

In [0]:
grid_search = GridSearchCV(model, parameters, cv=3)

***Hyperparamaters tunning for GradientBoostingClassifier***

In [0]:
from sklearn.ensemble import GradientBoostingClassifier


In [0]:
model = GradientBoostingClassifier()
parameters = parameters = {'n_neighbors': np.arange(1000, 1500,100)}
grid_search = GridSearchCV(model,parameters, cv= 3)